### PR TITLE
Fix method descriptor for `X extends Y`

### PR DESCRIPTION
**Pull request by rcoh**
Fixes: https://github.com/duchess-rs/duchess/issues/42
Fixes: https://github.com/duchess-rs/duchess/issues/149

Generics were not code generated properly. They should actually refer to the first class in the list of extending classes. This introduces `GenericsScope` to keep track of where generics were defined when creating descriptors.

### CODE REVIEW

File: macro/src/class_info.rs [hunk 1]
```diff
@@ -280,12 +280,13 @@ impl Constructor {
         }
     }
 
-    pub fn descriptor(&self) -> String {
+    pub fn descriptor(&self, ctx: &GenericsScope) -> String {
```
Comment (nikomatsakis) [inline macro/src/class_info.rs:283]:
```suggestion
    /// Returns the JVM descriptor script for the constructor.
    ///
    /// # Parameters
    ///
    /// * `ctx` is the generics scope of the class.
    pub fn descriptor(&self, ctx: &GenericsScope<'_>) -> String {
```

File: macro/src/class_info.rs [hunk 2]
```diff
@@ -317,16 +318,17 @@ impl Method {
...
```
Comment (nikomatsakis) [inline macro/src/class_info.rs:321]:
```suggestion
    /// Returns the JVM descriptor for the method.
    ///
    /// # Parameters
    ///
    /// * `ctx` is the generics scope of the class.
    pub fn descriptor(&self, ctx: &GenericsScope<'_>) -> String {
```

File: macro/src/class_info.rs [hunk 3]
```diff
@@ -424,8 +426,40 @@ impl Type {
         }
     }
 
-    pub fn descriptor(&self) -> String {
-        self.to_non_repeating().descriptor()
+    pub fn descriptor(&self, ctx: &GenericsScope) -> String {
```
Comment (nikomatsakis) [inline macro/src/class_info.rs:429]:
```suggestion
    /// Returns the JVM descriptor for this type, suitable for embedding a method descriptor.
    ///
    /// # Parameters
    ///
    /// * `ctx` is the generics scope where the type appears.
    pub fn descriptor(&self, ctx: &GenericsScope<'_>) -> String {
```

File: macro/src/class_info.rs [hunk 4]
```diff
@@ -437,18 +471,24 @@ pub enum NonRepeatingType {
 }
 
 impl NonRepeatingType {
-    pub fn descriptor(&self) -> String {
+    pub fn descriptor(&self, ctx: &GenericsScope) -> String {
```
Comment (nikomatsakis) [inline macro/src/class_info.rs:474]:
```suggestion
    /// Returns the JVM descriptor for this type, suitable for embedding a method descriptor.
    ///
    /// # Parameters
    ///
    /// * `ctx` is the generics scope where the type appears.
    pub fn descriptor(&self, ctx: &GenericsScope<'_>) -> String {
```

File: macro/src/class_info.rs [hunk 5]
```diff
@@ -424,8 +426,40 @@ impl Type {
...
+    /// Add an additional layer to this scope (e.g. combining generics from a class with generics from a method)
+    ///
+    /// The newly provided generics are higher priority than the inner generics (but
+    /// I don't think we can have namespace collisions here in Java anyway)
+    fn nest(&'a self, generics: &'a [Generic]) -> GenericsScope<'a> {
+        GenericsScope::Generics(generics, self)
+    }
+}
+
+impl<'a> From<&'a ClassInfo> for GenericsScope<'a> {
```
Comment (nikomatsakis) [inline macro/src/class_info.rs:460]: Why not add a method like `class_info.generics_scope()`?

Seems a bit more obvious to me.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
I'd prefer not to have the `From` impl.

File: macro/src/class_info.rs [hunk 6]
```diff
@@ -437,18 +471,24 @@ pub enum NonRepeatingType {
...
-    pub fn descriptor(&self) -> String {
+    pub fn descriptor(&self, ctx: &GenericsScope) -> String {
         match self {
             NonRepeatingType::Ref(r) => match r {
                 RefType::Class(c) => format!("L{};", c.name.to_jni_name()),
-                RefType::Array(r) => format!("[{}", r.descriptor()),
+                RefType::Array(r) => format!("[{}", r.descriptor(ctx)),
 
                 // FIXME(#42): The descriptor actually depends on how the type
                 // parameter was declared!
```
Comment (nikomatsakis) [inline macro/src/class_info.rs:481]: Oh, and you can remove this FIXME, and add `Closes #42` to the list :)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

**Comment by nikomatsakis** [comment]
Nice job! Left some nits.

File: macro/src/class_info.rs (hunk 5 above)
Comment (rcoh) [reply macro/src/class_info.rs:460]: +1 that's much clearer

File: macro/src/class_info.rs [hunk 7]
```diff
@@ -172,6 +172,10 @@ impl ClassInfo {
             | (Privacy::Default, ClassKind::Class) => false,
         }
     }
+
+    pub fn generics_scope(&self) -> GenericsScope {
```
Comment (nikomatsakis) [inline macro/src/class_info.rs:176]:
```suggestion
    pub fn generics_scope(&self) -> GenericsScope<'_> {
```

This is the case I find most important! Got to signal that the return is tied to `self`

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**